### DISCUSSION TITLE
TLS session storage, memory + permanent

### DISCUSSION THREAD

**Discussion by icing** [curl/curl#15684] [General]
Thinking about storing TLS sessions more permanently, e.g. in a file. This would make these available the next time you start curl. Reusing them is beneficial for performance, especially now that we add 0RTT support.

### Current State

Curl's TLS session cache currently just sits in memory and uses a plethora of criteria when looking up sessions for reuse: `hostname`+`port`, `transport`, `scheme`, `connect_to hostname`+`port`, `verifypeer`, `verifyhost`, `verifystatus`, `version`, `version_max`, `ssl_options`, `cert_blob`, `ca_info_blob`, `issuercert_blob`, `CApath`, `CAfile`, `issuercert`, `clientcert`, `username+password` (SRP), `cipher_list`, `cipher_list13`, `curves`, `CRLfile`, `pinned_key`.

Wow. Only when *all* these criteria match is a TLS session re-used in a new connection.

### Storing a subset

We cannot do this exactly the same in a permanent session store. For example, SRP `username+password` we definitely do **not** want to store. Same for `clientcert`. Which means we'll store only a *subset* of the memory TLS session cache. 

Which should be fine, it's a *cache* anyway. Transfers using client certificates or SRP, will not benefit from the *permanent* cache. But the *memory* cache would continue to work for them.

#### Unverified

Sessions from a TLS connection that have not verified the peer (`verifypeer` or `verifyhost` disabled) should be disregarded as well. Such connections have no long-term security properties. They encrypt the traffic, but everything else is untrustworthy.

#### `verifystatus` and `CRLfile`

For transfers that are concerned about certificate revocation, sessions from a permanent cache seem counter productive. A TLS session resumptions does not inspect the certificate again and revocations will not be detected.

I propose to exclude sessions from such connections from permanent storage.

#### `pinned_key`

Pinned keys seem to be a thing of the past. It seems not worth it to bother a permanent cache with sessions from such connections. 

#### Verified and `connect_to_*`

Dealing with sessions from peer-verified connections means that `connect_to_hostname` and`connect_to_port` are irrelevant. It does not matter if we got the session through an SSH tunnel, proxy tunnel or directly (the memory cache could also change its logic here).

For storing verified session, we can disregard connect_to_*.

### Verified how?

To re-use a verified TLS session, it is important that the TLS trust anchors are the same as before. This means that 
`cert_blob`, `ca_info_blob`, `issuercert_blob`, `CApath`, `CAfile`, `issuercert` are still the same. This gets tricky.

The blobs and file contents can be huge and in the case of `CApath` it can be a large set of files. A *exact* match of all these seems excessive. 

True, in the memory cache, we do not match the *contents* of `CApath` and `CAfile` either. However, these file have been parsed at application start and, for most TLS backends, will not be parsed again. So matching file names seems safe. When a session has been saved yesterday, the contents of those files might have changed. (Also, file/path names can be relative and may not point to the same location in the file system.)

We could exclude such sessions also, however I suspect that there will be many use cases with internal CAs that would like the benefit from session reuse.

If we *do* include these sessions for permanent storage, we need to clarify how we store this information. It seems impractical to store the blobs verbatim. An alternative is to store a SHA256 hash of the information. This we'll then need to use in memory cache matching as well. Otherwise reuse would never happen.

**Issue**:  how to handle this? Another CURLOPT_SSL to allow/forbid it? Only exclude sessions from relative file/path names?

#### `transport` and `scheme`

`transport` needs to be stored in a permanent cache as servers will most likely have different handling. For example, the "max early data" setting in TLS sessions will often differ between HTTP/3 and HTTP/2 endpoints.

`scheme` seems of limited value. Maybe it should be removed from the memory cache as well. In the age of ALPN protocol negotiation, there seems to be no need for remembering the scheme???

### Properties for permanent storage

This leaves the following criteria for "long-term" session storage:

* must have used verifypeer + verifyhost
* must not have used  `verifystatus`, `CRLfile` or `pinned_key`
* must not have used `cipher_list`, `cipher_list13` or `curves`

What a permanent TLS cache will store then is
* TLS implementation name (curl invocation may use different TLS stack)
* timestamp session was received
* lifetime seconds (where available)
* peer `hostname` + `port`
* `transport` (TCP or QUIC)
* SSL `version`, `version_max`, `ssl_options`
* SSL `cipher_list`, `cipher_list13`, `curves`
* SHA256 of `cert_blob`, `ca_info_blob`, `issuercert_blob`, `CApath`, `CAfile`, `issuercert` ???
* Session blob itself
* ALPN negotiated (Optional)
* max early data (0 when not applicable)
* QUIC transport params blob (Optional)

**Update**: added received timestamp and lifetime to store

**Comment by dfandrich** [comment]
This feature should make a huge difference in a lot of workloads!

> The blobs and file contents can be huge and in the case of CApath it can be a
> large set of files. A exact match of all these seems excessive.

In the case of individual certificates, storing the fingerprint should be
sufficient.

> True, in the memory cache, we do not match the contents of CApath and CAfile
> either. However, these file have been parsed at application start and, for most
> TLS backends, will not be parsed again. So matching file names seems safe. When
> a session has been saved yesterday, the contents of those files might have
> changed. (Also, file/path names can be relative and may not point to the same
> location in the file system.)

Isn't there nowadays code that parses the entire CA bundle the first time it's
used? It could be hashed at that time, or all the fingerprints could be hashed
with almost no penalty. But doing that on a CA path would cause a huge slowdown
the first time, so not really an option.  If we don't want to do that, then the
path is about all we have. Unless we include in the cache the fingerprint of
only the CA certificates used to verify the server cert. Relative paths can be
made absolute for storage in the cache.

> scheme seems of limited value. Maybe it should be removed from the memory cache
> as well. In the age of ALPN protocol negotiation, there seems to be no need for
> remembering the scheme???

Isn't ALPN done before certificate validation? So, couldn't there be different
certificates presented depending on the negotiated protocol?

> What a permanent TLS cache will store then is
> 
>   • TLS implementation name (curl invocation may use different TLS stack)

The exact version number should really be stored as well, since I'm assuming
the session data will be internal to the TLS stack and therefore could change
from version to version. Unless we can guarantee that all TLS stacks we will
ever support store their own version number in their opaque blobs and will
always do this checking themselves.

Is it worthwhile warning the user if the session file has insecure permissions,
à la ssh? Or, will there actually be nothing private stored in it? Is there
enough information in the file that an attacker could use it to resume a
session that was established using a private certificate without having that
certificate? Things like host names will be stored in there that should often
be private, but curl doesn't warn about .netrc being world-readable or
world-writable now either, so this would be a new feature.

Another thing to consider is cache pruning. Presumably, the format will be curl
specific and won't necessarily be easy for the user to manually clean out
periodically.  How will entries be removed?  Only when they are used but fail?
Perhaps the date they were added to the cache should be included in each entry
and those more than X days (hours?) old are automatically removed each time the
file is rewritten, regardless of if they have been used recently or not.

**Reply by icing** [reply]
>> The blobs and file contents can be huge and in the case of CApath it can be a large set of files. A exact match of all these seems excessive.
> In the case of individual certificates, storing the fingerprint should be sufficient.

I am not sure how that could be achieved. Do you have a suggestion?

> Isn't there nowadays code that parses the entire CA bundle the first time it's used? It could be hashed at that time, or all the fingerprints could be hashed with almost no penalty. But doing that on a CA path would cause a huge slowdown the first time, so not really an option. If we don't want to do that, then the path is about all we have. Unless we include in the cache the fingerprint of only the CA certificates used to verify the server cert. Relative paths can be made absolute for storage in the cache.

We do not parse the CA file ourself. That is up to the TLS backend and I am not sure if all allow iterating over the results.

>> scheme seems of limited value. Maybe it should be removed from the memory cache as well. In the age of ALPN protocol negotiation, there seems to be no need for remembering the scheme???
> Isn't ALPN done before certificate validation? So, couldn't there be different certificates presented depending on the negotiated protocol?

Theoretically, yes. Anything in the ClientHello could lead to a different certificate selection on the server. It could also vary it randomly if it wants. But the server then would also need some careful code around TLS session resumption. For the client, this should not matter, as it verifies the peer only for the hostname.

>> What a permanent TLS cache will store then is • TLS implementation name (curl invocation may use different TLS stack)
> The exact version number should really be stored as well, since I'm assuming the session data will be internal to the TLS stack and therefore could change from version to version. Unless we can guarantee that all TLS stacks we will ever support store their own version number in their opaque blobs and will always do this checking themselves. 

We might need that, but I'd hope the TLS stacks take care of that situation.

> Is it worthwhile warning the user if the session file has insecure permissions, à la ssh? Or, will there actually be nothing private stored in it? Is there enough information in the file that an attacker could use it to resume a session that was established using a private certificate without having that certificate? 

That is why I think we should not store session with SRP or client certificates. At least with 0RTT there is a chance to exploit such a session, depending on how the server protects itself.

> Things like host names will be stored in there that should often be private, but curl doesn't warn about .netrc being world-readable or world-writable now either, so this would be a new feature. 

As @vszakats wrote, it might be a good idea to hash the hostnames like SSH does.

> Another thing to consider is cache pruning. Presumably, the format will be curl specific and won't necessarily be easy for the user to manually clean out periodically. How will entries be removed? Only when they are used but fail? 

I would think a text file, with a session per line, blobs base64 encoded would work best here. It's nothing we need random access to, I think.

> Perhaps the date they were added to the cache should be included in each entry and those more than X days (hours?) old are automatically removed each time the file is rewritten, regardless of if they have been used recently or not.

Yes, I forgot the "session received" timestamp that needs to be stored. In TLSv1.3 tickets carry also the "seconds this session is supposed to be valid" information, but not all TLS stacks give access to that. (GnuTLS does not, OpenSSL does)

**Reply by bagder** [reply]
I'm +1 for hashing the hostname + port to reduce privacy leakage risk. Also, +1 on avoiding SRP and client certificates in the data for security's sake.

I don't want us to add a warning for insecure permissions because of general warning fatigue - I suspect it does nothing good at all and just gives us a false impression that we try to do something good. But I'm willing to listen to what others think and see if there's some lessons on this drawn from for example OpenSSH.

What is the worst thing that can be achieved if you get someone else's TLS session and figure out what servers it covers?

**Reply by icing** [reply]
See #15707 for a stab at this: the complete SSL parameters are part of a SHA256 hash, except the sensitive values client cert and SRP user/pass. This hash is used for lookups and storing. If we persist it on a file, you'd need to know all names and SSL parameters to reproduce the hash.

If we store this, someone could check if you visited google.com. That is leaking. But an attacker cannot deduce hosts they do not know.

There is a very small chance that this session pool ever has a hash collision and would hand out a SSL session for a connection it does not belong to. But then a server will not resume it and start the TLS handshake all over again, AFAICT.

**Reply by bagder** [reply]
> If we store this, someone could check if you visited google.com. That is leaking. But an attacker cannot deduce hosts they do not know.

And if we add a salt in there, not even that should be possible, right? I mean we could just add a random number not used for anything else, to prevent the hash from being static for a given situation.

**Reply by icing** [reply]
But how would we use it then? If we use a random salt for each session, it is much more difficult to find a certain hostname in a session file, yes. But it is also more expensive for curl to lookup a session for a host at runtime.

Because it then would have to try all present salts, right?

**Reply by bagder** [reply]
It would need to store the salt in a readable way next to the hash, like in `/etc/passwd`: [salt]:[hash] - it prevents a fixed hash in the file to represent a known host + config.

A common TLS config against `google.com` might otherwise always generate for example `f0VMRgIBAQAAAAAAAAA`, and we could just grep for that and know the host.

Admittedly, that's the only added "protection" we get with the salt. A user can also just do the hash itself and then check for that, but then at least it needs to do that for every line in the session store.

**Reply by icing** [reply]
Agreed. But that would mean that when libcurl looks for an existing session to `google.com` it would need to try all salts from all sessions in the pool. It's doable at O(n) cost.

**Reply by bagder** [reply]
Ah yes, good point... Can we do it with a single salt for the whole file? That still removes the fixed hash but makes it easier to find matches.

**Reply by icing** [reply]
I am now pondering to use an "internal id", the SHA256 of all relevant parameters, and an "external id", which is a salted HMAC of the internal id. The external id we can store in a file which has then similar properties to SSH's `known_hosts`. It could then work as follows.

1. Loading such a file, all entries in the session cache have only the external id, not the internal one. 
2. The connection filter wanting a session has its internal id and asks the session cache.
3. The session cache does now find an entry via:
    a. looking for an entry that has the same internal id
    b. If there is none, it iterates over all entries with unknown internal id and compares the external id (using the salt to compute the HMAC from the lookup internal id).
    c. If the entry external id matches, it stores a the internal ID in the entry. The next lookup of this id will be fast
    d. If no entry matches, lookup is unsuccessful. 

Over time, the lookup performance improves. This should work for small session caches. But applications that want a large cache would need to work differently. For those, libcurl would need to provide an interface for session lookup and storage using the internal id, so the app can use a database or something along those lines, not reading in all sessions at start.

But let's focused on the command line curl use first before we design for applications we do not know.

**Reply by jay** [reply]
> The exact version number should really be stored as well, since I'm assuming the session data will be internal to the TLS stack and therefore could change from version to version. Unless we can guarantee that all TLS stacks we will ever support store their own version number in their opaque blobs and will always do this checking themselves.

This seems like a good point to consider the uniqueness of curl. There are users with multiple builds, versions and SSL libraries.

**Reply by icing** [reply]
> > The exact version number should really be stored as well, since I'm assuming the session data will be internal to the TLS stack and therefore could change from version to version. Unless we can guarantee that all TLS stacks we will ever support store their own version number in their opaque blobs and will always do this checking themselves.
> 
> This seems like a good point to consider the uniqueness of curl. There are users with multiple builds, versions and SSL libraries.

We definitely need to store the TLS implementation together with each session.

**Reply by jay** [reply]
I'm saying that more broadly different builds of curl should not share sessions because over time the implementation may vary. A quick way to handle this would probably be hash curl_version

**Comment by vszakats** [comment]
To check a matching SRP `username+password` and `clientcert` it
may be an option to store a salted secure hash of these values. The
secure hash needs to be fast enough to not cancel the benefit of
caching.

Such method could be used to hide the `host:port` value too, like it's
done in SSH `known_hosts`.

Without revealing a cleartext `host:port`, the session blob alone is
less useful for an attacker.

**Reply by vszakats** [reply]
One further option is to accept a session file passphrase from the command-line and save/load the disk file with symmetric encryption. (Works in addition to the above too.)

**Reply by icing** [reply]
The hash values in ` known_hosts` protect against someone finding out what you connect to. If you want to find an entry for a particular host, `ssh-keygen -F` will do that for you. It may be of value in a curl session file to hide the host names as well.

But.

If we store sessions with user authentication (SRP or clientcert), the threat scenario is different from `known_hosts`. Because you cannot do much with a SSH host key, but you could use an authenticated TLS session to try to subvert the server. At least when it also allows early data and is not careful about it.

That is why I am hesitant in storing such sessions.

**Reply by vszakats** [reply]
Agreed, it's safer to not include auth info at all. And do the hashing for the `host:port`.

**Reply by icing** [reply]
> One further option is to accept a session file passphrase from the command-line and save/load the disk file with symmetric encryption. (Works in addition to the above too.)

Doable and maybe a good idea if we store real secret stuff in there. However, I would like to avoid that. Maybe it's good to talk about the goals of this feature a bit.

For me, the session file is there to give performance improvements. People enable it in their `.curlrc` file and then forget about it. It will give them faster transfers when a previous session can be reused. But curl will work the same without it.

From a security point of view, and maybe someone who is more expert than me should have a look, I would only store TLSv1.3 session tickets there. TLSv1.2 session ids and tickets I do not know very well, and I read there were issues in the past.

So, speaking about TLSv1.3 tickets: as I understand them, they contain a pre-shared-key (PSK) for resumption and that is also used to encrypt 0RTT data. The TLS handshake is still performed with the only difference that the server sends proof it knows its side of the PSK and no longer sends the certificate. Otherwise, it's a full handshake with newly negotiated keys and the PSK will no longer be used after that.

Without 0RTT and without any client authentication in the TLS handshake, obtaining a copy of the session ticket seems should not be of any value to an attacker (Note: my guess, expert opinions welcome).

With 0RTT, the situation differs: there is no forward secrecy for 0RTT data. Someone in possession of the ticket can decrypt it (but only the 0RTT data, not the rest of the connection). 

It might be a good idea to clear a session ticket from the cache when 0RTT data is sent. Also important to note: we do not send 0RTT data unless the `CURLSSLOPT_EARLYDATA` is set for a transfer (e.g. --tls-earlydata is given to curl).

**Comment by icing** [comment]
Summarising the Status Quo:

In #15707 we now have the following:

We use `ssl_peer_key`s internally for session lookup and storage. These are generated for the `peer` a TLS connection filter (or QUIC connection filter) talks to. They carry the peer+port and all relevant TLS configuration (excluding cliencert and SRP auth). The key is for the filter instance, meaning a proxy TLS filter will have a different key than the TLS filter that talks to the origin server - as should be because they talk to different peers.

Examples: 
* `SESS:one.http.curl.se:64664:NO-VRFY-PEER:NO-VRFY-HOST:TLSOPT-40:CA-/etc/ssl/cert.pem` is the peer key in a test case where host+peer verification is off and sending EARLYDATA option is enabled
* `SESS:one.http.curl.se:64778:CA-/Users/sei/projects/curl/tests/http/gen/ca/ca.rsa2048.cert.pem` is the peer key in a test case where a local CA file is given with relative path, made absolute.

When a client certificate or SRP auth is configured, those are remembered in the session cache separately. This ensures that such a session will only be handed out to filters that have the same configuration. And vice versa: a filter will not reuse a session created with auth information.

Export:
Exported sessions will get a second "id" that is the randomly salted HMAC_SHA256 of the peer key (together). The libcurl application can then decide which id to use when storing sessions. If the application has private storage, it may use the peer key. If it wants to use a file, like the curl command line would, it stores the HMAC and not the peer key. 

Since each HMAC carries a unique salt, such a session file would not allow "grepping" for known entries.

Import:
When importing sessions into libcurl, the application will either provide the peer key or the HMAC. If only the HMAC is imported, the peer key is unknown. Lookups into the cache always use a peer key, however and when no entry matches, the entries with HMAC and unknown peer keys are tested. If one matches, the peer key for the entry becomes known and will be remembered for future lookups.


Session Data itself:
Together with the id/key, libcurl will export the session data itself as chunk of bytes. This is supported by all TLS backends where we used the session cache except Schannel. Schannel stores "Credential" objects which we cannot persist in and read back from a file. I split the session cache API so that it knows when objects are cached, so it will  not export those entries.

TODO
- storing TLS backend + version in session cache and using that also for lookup
- storing timestamp when adding sessions and lifetime seconds where available (not all TLS backends support this)
- API design for import/export of sessions in libcurl
- curl implementation for file storage

**Comment by icing** [comment]
Rework of TLS session ticket handling in #15774. Time to design how we do im-/export.

**Comment by icing** [comment]
I propose the following additions for import/export of TLS sessions:

```c
CURL_EXTERN CURLcode
curl_share_ssl_session_import(CURLSH *share, const char *session_key,
                              const unsigned char *shmac, size_t shmac_len,
                              const unsigned char *sdata, size_t sdata_len);

typedef CURLcode curl_ssl_export_function(void *userptr,
                                          const char *session_key,
                                          const unsigned char *shmac,
                                          size_t shmac_len,
                                          const unsigned char *sdata,
                                          size_t sdata_len,
                                          size_t sdata_len,
                                          curl_off_t valid_until,
                                          int ietf_tls_id,
                                          char *alpn,
                                          size_t earlydata_max);

CURL_EXTERN CURLcode
curl_share_ssl_session_export(CURLSH *share,
                              curl_ssl_export_function *export_fn,
                              void *userptr);
```

with `curl_share_ssl_session_import()` either getting the `session_key` *or* the `shmac`. The export function can choose what to store. It can return `CURL_OK` when done or `CURLE_AGAIN` to keep on exporting.

How does that sound?

**Update:** added informative paramter to export function to allow caller to decide if they are interested. These params are informative as they are not necessary for importing the session again.

**Reply by bagder** [reply]
I think it sounds good and a feasible API. So the export method then calls the export callback once per existing session to save?

**Reply by icing** [reply]
Yes.

**Reply by dfandrich** [reply]
I'm a bit worried about forward compatibility with this API. If TLS 1.4 adds something new that the callback should get, or we decide to make more information available (# of times each session was used?) we're stuck. Some of the options:
1. Ignore it. If this happens, create a curl_share_ssl_session_import2() with a new callback signature.
2. Add another argument to the callback `int export_version` set to 1 in this implementation. If ever we need to add more, bump it to 2 and just add more arguments to the end of the function signature and have the callback check the version number to see if it can access those new arguments or not. This should work technically in C, but I'm not 100% sure that it doesn't depend on undefined behaviour.
3. A more libcurl-like API would be to revert to the previous callback signature providing just the mandatory information and create a new libcurl function `CURLcode curl_session_getinfo(const char *session_key, CURLSESSIONINFO info, ...)` or similar that would return one piece of additional information about the session. Callbacks that don't care about the info wouldn't call it but those that do could get an extensible list of whatever additional info they needed in a forward-compatible manner.

**Reply by icing** [reply]
It is always hard to anticipate the unexpected. =)

Note that all the parameters passed here, outside key+shmac+sdata are informative only and need not be stored. I covered this in the [man page](https://github.com/curl/curl/blob/a8661e59de6a5fb9bf9115d0e5c4c29707ba9416/docs/libcurl/curl_easy_ssls_export.md).

Worth noticing here: the `sdata` passed to the export callback is not simply the stuff from the TLS backend. It is more. I created a "pack/unpack" of session information pieces. See [vtls_spack.c](https://github.com/curl/curl/pull/15905/files#diff-ae993d46e6562c429fb32ee2401bf0893f66527fe51d015c1a147f8d672970c5).

This means, when TLS v1.4 comes with something new we'd like to keep, we can add it to the session export without changing the API. It is just not available as criteria to the export caller.

I think. once there is some additional data to be made visible, we can discuss if an "export2" or a "session info" is the better way to extend this.

**Comment by dfandrich** [comment]
Can the callback get information about a session to determine whether it
wants to store it? For example, I would expect there to be applications that
only want to store sessions to a specific host, or to only hosts in a given
domain, or only sessions that expire more than X minutes from now, or perhaps
even other criteria. For example, a crawler that visits 1E+6 sites and phones
home about them probably doesn't care about sessions to any site but home and
it would be wasteful to store (and restore) all the rest knowing that it's
unlikely to ever use them.

**Reply by icing** [reply]
Excellent point. I added "informative" parameters to the export function. These do not need to be stored by the caller for later import.

As for filtering on destinations, the function gets the `session_key` which starts with `hostname:port`. We could document that, but I'd prefer to leave the rest of the key undocumented.

**Comment by yedayak** [comment]
Is there a reason the option was called `--ssl-sessions` in https://github.com/curl/curl/pull/15924? Compare to `--tls-sessions` for example?
It's a bit confusing, since this is especially useful in conjunction with`--tls-earlydata`, which has `tls` in the name.

**Reply by icing** [reply]
> Is there a reason the option was called `--ssl-sessions` in #15924? Compare to `--tls-sessions` for example? It's a bit confusing, since this is especially useful in conjunction with`--tls-earlydata`, which has `tls` in the name.

The terms `ssl` and `tls` have been a source of confusion  from the very beginning - unrelated to curl. We cannot fix that. I think the best we can do is to use them consistently in curl. Since we cannot rename all existing `ssl` things to `tls` for backward compatibility, our options are limited.

I think we should use `tls` for everything related to TLSv1.3 and newer. And use `ssl` for things across all versions. Along those lines `--tls-earlydata` is a feature for TLSv1.3 only, while `--ssl-sessions` works for all versions.

Naming is hard.
